Can't install next-auth when next 13 is installed

### Environment

  System:
    OS: Linux 5.4 Ubuntu 22.04.1 LTS 22.04.1 LTS (Jammy Jellyfish)
    CPU: (4) x64 Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
    Memory: 3.32 GB / 7.77 GB
    Container: No
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 18.10.0 - /usr/local/share/nvm/current/bin/node
    Yarn: 1.22.19 - /usr/bin/yarn
    npm: 8.19.2 - /usr/local/share/nvm/current/bin/npm
  npmPackages:
    next: ^13.0.1 => 13.0.1 
    react: ^18.2.0 => 18.2.0 

### Reproduction URL

N/A

### Describe the issue

Installation of `next-auth` fails when next v13.0.1 is already installed.
```bash
npm ERR! Found: next@13.0.1
npm ERR! node_modules/next
npm ERR!   next@"^13.0.1" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer next@"^12.2.5" from next-auth@4.15.0
npm ERR! node_modules/next-auth
npm ERR!   next-auth@"*" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

### How to reproduce

Run `npm i next-auth` in a project with next 13.0.1 already installed.

### Expected behavior

Should install.

--- a/apps/dev/components/footer.js
+++ b/apps/dev/components/footer.js
@@ -17,9 +17,7 @@ export default function Footer() {
           <a href="https://github.com/nextauthjs/next-auth-example">GitHub</a>
         </li>
         <li className={styles.navItem}>
-          <Link href="/policy">
-            <a>Policy</a>
-          </Link>
+          <Link href="/policy">Policy</Link>
         </li>
         <li className={styles.navItem}>
           <em>{packageJSON.version}</em>

--- a/apps/dev/components/header.js
+++ b/apps/dev/components/header.js
@@ -64,49 +64,31 @@ export default function Header() {
       <nav>
         <ul className={styles.navItems}>
           <li className={styles.navItem}>
-            <Link href="/">
-              <a>Home</a>
-            </Link>
+            <Link href="/">Home</Link>
           </li>
           <li className={styles.navItem}>
-            <Link href="/client">
-              <a>Client</a>
-            </Link>
+            <Link href="/client">Client</Link>
           </li>
           <li className={styles.navItem}>
-            <Link href="/server">
-              <a>Server</a>
-            </Link>
+            <Link href="/server">Server</Link>
           </li>
           <li className={styles.navItem}>
-            <Link href="/protected">
-              <a>Protected</a>
-            </Link>
+            <Link href="/protected">Protected</Link>
           </li>
           <li className={styles.navItem}>
-            <Link href="/protected-ssr">
-              <a>Protected(SSR)</a>
-            </Link>
+            <Link href="/protected-ssr">Protected(SSR)</Link>
           </li>
           <li className={styles.navItem}>
-            <Link href="/api-example">
-              <a>API</a>
-            </Link>
+            <Link href="/api-example">API</Link>
           </li>
           <li className={styles.navItem}>
-            <Link href="/credentials">
-              <a>Credentials</a>
-            </Link>
+            <Link href="/credentials">Credentials</Link>
           </li>
           <li className={styles.navItem}>
-            <Link href="/email">
-              <a>Email</a>
-            </Link>
+            <Link href="/email">Email</Link>
           </li>
           <li className={styles.navItem}>
-            <Link href="/middleware-protected">
-              <a>Middleware protected</a>
-            </Link>
+            <Link href="/middleware-protected">Middleware protected</Link>
           </li>
         </ul>
       </nav>

--- a/apps/dev/package.json
+++ b/apps/dev/package.json
@@ -19,7 +19,7 @@
     "@next-auth/typeorm-legacy-adapter": "workspace:*",
     "@prisma/client": "^3",
     "faunadb": "^4",
-    "next": "12.3.1",
+    "next": "13.0.1",
     "next-auth": "workspace:*",
     "nodemailer": "^6",
     "react": "^18",

--- a/packages/next-auth/package.json
+++ b/packages/next-auth/package.json
@@ -120,7 +120,7 @@
     "jest-environment-jsdom": "^28.1.1",
     "jest-watch-typeahead": "^1.1.0",
     "msw": "^0.42.3",
-    "next": "12.3.1",
+    "next": "13.0.1",
     "postcss": "^8.4.14",
     "postcss-cli": "^9.1.0",
     "postcss-nested": "^5.0.6",

--- a/packages/next-auth/src/core/lib/cookie.ts
+++ b/packages/next-auth/src/core/lib/cookie.ts
@@ -1,6 +1,7 @@
-import type { IncomingHttpHeaders } from "http"
 import type { CookiesOptions } from "../.."
 import type { CookieOption, LoggerInstance, SessionStrategy } from "../types"
+import type { NextRequest } from "next/server"
+import type { NextApiRequest } from "next"
 
 // Uncomment to recalculate the estimated size
 // of an empty session cookie
@@ -128,10 +129,10 @@ export class SessionStore {
 
   constructor(
     option: CookieOption,
-    req: {
-      cookies?: Partial<Record<string, string> | Map<string, string>>
-      headers?: Headers | IncomingHttpHeaders | Record<string, string>
-    },
+    req: Partial<{
+      cookies: NextRequest["cookies"] | NextApiRequest["cookies"]
+      headers: NextRequest["headers"] | NextApiRequest["headers"]
+    }>,
     logger: LoggerInstance | Console
   ) {
     this.#logger = logger
@@ -140,7 +141,14 @@ export class SessionStore {
     const { cookies } = req
     const { name: cookieName } = option
 
-    if (cookies instanceof Map) {
+    if (typeof cookies?.getAll === "function") {
+      // Next.js ^v13.0.1 (Edge Env)
+      for (const { name, value } of cookies.getAll()) {
+        if (name.startsWith(cookieName)) {
+          this.#chunks[name] = value
+        }
+      }
+    } else if (cookies instanceof Map) {
       for (const name of cookies.keys()) {
         if (name.startsWith(cookieName)) this.#chunks[name] = cookies.get(name)
       }

--- a/packages/next-auth/src/next/middleware.ts
+++ b/packages/next-auth/src/next/middleware.ts
@@ -184,7 +184,7 @@ export type WithAuthArgs =
  * [Documentation](https://next-auth.js.org/configuration/nextjs#middleware)
  */
 export function withAuth(...args: WithAuthArgs) {
-  if (!args.length || args[0] instanceof NextRequest) {
+  if (!args.length || args[0] instanceof Request) {
     // @ts-expect-error
     return handleMiddleware(...args)
   }

